blocks/edit/prose/loc/loc-utils.js

```diff
@@ -0,0 +1,946 @@
+import {
+  DOMSerializer,
+  Fragment,
+  Slice,
+} from 'da-y-wrapper';
+import getSheet from '../../../shared/sheet.js';
+import { htmlDiff } from './htmldiff.js';
+
+// Constants
+const LOC_COLORS = {
+  UPSTREAM: 'rgba(70, 130, 180, 0.2)',
+  LOCAL: 'rgba(144, 42, 222, 0.2)',
+  DIFF: 'rgba(150, 150, 150, 0.1)',
+};
+
+const LOC_TEXT = {
+  UPSTREAM: 'Upstream',
+  LOCAL: 'Local',
+  DIFF: 'Difference',
+};
+
+// DOM Creation Utilities
+function createElement(tag, className = '', attributes = {}) {
+  const element = document.createElement(tag);
+  if (className) element.className = className;
+  Object.entries(attributes).forEach(([key, value]) => {
+    element.setAttribute(key, value);
+  });
+  return element;
+}
+
+function createTooltip(text) {
+  const tooltip = createElement('span', 'loc-tooltip');
+  tooltip.textContent = text;
+  return tooltip;
+}
+
+function createButton(className, type = 'button', attributes = {}) {
+  const button = createElement('button', className, { type, ...attributes });
+  return button;
+}
+
+// Transaction Utilities
+function createContentTransaction(view, startPos, endPos, filteredContent) {
+  const { tr } = view.state;
+
+  if (filteredContent.length > 0) {
+    const newFragment = Fragment.fromArray(filteredContent);
+    const newSlice = new Slice(newFragment, 1, 1);
+    return tr.replace(startPos, endPos, newSlice);
+  }
+
+  return tr.delete(startPos, endPos);
+}
+
+function isLocNode(node) {
+  return node?.type?.name === 'loc_deleted' || node?.type?.name === 'loc_added';
+}
+
+function isValidPosition(pos) {
+  return pos !== null && pos !== undefined;
+}
+
+// Content Filtering Utilities
+function filterNodeContent(node) {
+  if (!node?.content?.content) {
+    return [];
+  }
+
+  return node.content.content.filter((child) => {
+    // Handle text nodes
+    if (child.type.name === 'text') {
+      return child.text?.trim().length > 0;
+    }
+
+    // Handle nodes with content
+    if (child.content) {
+      // Check if it has meaningful content
+      if (child.content?.content?.length > 0) {
+        return true;
+      }
+      // For nodes without nested content, check if they have any attributes or marks
+      return child.attrs || child.marks?.length > 0;
+    }
+
+    // For other node types, assume they're meaningful unless explicitly empty
+    return true;
+  });
+}
+
```

Comment:
Utility methods could be shared with others (I bet you have a createElement somewhere else). Externalize and... tests.